subsys/senss/sensor_mgmt.c

```diff
@@ -0,0 +1,413 @@
+		LOG_ERR("malloc memory for struct senss_connection error");
+		return SENSS_SENSOR_INVALID_HANDLE;
+	}
+	conn->data = (uint8_t *)conn + sizeof(*conn);
+
+	/* create connection for reporter to application(client = NULL) */
+	ret = init_each_connection(ctx, conn, reporter, NULL);
+	if (ret) {
+		LOG_ERR("%s, init_each_connection error:%d", __func__, ret);
+		free(conn);
+		return SENSS_SENSOR_INVALID_HANDLE;
+	}
+
+	ret = senss_bind_conn(ctx, conn);
+	if (ret) {
+		LOG_ERR("%s, senss_bind_conn error:%d", __func__, ret);
+		free(conn);
+		return SENSS_SENSOR_INVALID_HANDLE;
+	}
+
+	LOG_INF("%s ready: %s, state:0x%x, conn:%d",
+			__func__, reporter->dev->name, reporter->state, conn->index);
+
+	return conn->index;
+}
+
+int close_sensor(struct senss_connection *conn)
+{
+	struct senss_mgmt_context *ctx = get_senss_ctx();
+	int ret;
+
+	ret = senss_unbind_conn(ctx, conn);
+	if (ret) {
+		LOG_ERR("%s, senss_unbind_conn error:%d", __func__, ret);
+		return SENSS_SENSOR_INVALID_HANDLE;
+	}
+
+	/* reporter to application: free connection
+	 * connection between reporter and client derived from device tree: free conn data
+	 */
+	if (!conn->sink) {
+		free(conn);
+	} else {
+		free(conn->data);
+	}
+
+	return 0;
+}
+
+int senss_init(void)
+{
+	struct senss_mgmt_context *ctx = get_senss_ctx();
+	struct senss_sensor_dt_info *dt_info = &sensors_dt[0];
+	struct senss_sensor *sensor;
+	enum senss_sensor_state state;
+	int ret = 0;
+	int i;
+
+	LOG_INF("%s begin...", __func__);
+
+	if (ctx->senss_initialized) {
+		LOG_INF("senss is already initialized");
+		return 0;
+	}
+
+	ctx->sensor_num = ARRAY_SIZE(sensors_dt);
+	if (ctx->sensor_num == 0) {
+		LOG_WRN("no sensor created yet");
+		return 0;
+	}
+
+	for (i = 0; i < ctx->sensor_num; i++) {
+		sensor = create_sensor(dt_info);
+		if (!sensor) {
+			LOG_ERR("%s, create_sensor error", __func__);
+			return -EINVAL;
+		}
+		ctx->sensor_db[i] = sensor;
+		dt_info++;
+	}
+
+	sort_sensors(ctx);
+
+	for_each_sensor(ctx, i, sensor) {
+		ret = init_sensor(ctx, sensor);
+		if (ret) {
+			LOG_ERR("sensor:%s initial error", sensor->dev->name);
+		}
+		state = (ret ? SENSS_SENOSR_STATE_NOT_READY : SENSS_SENOSR_STATE_READY);
+		ret = set_sensor_state(sensor, state);
+		if (ret) {
+			LOG_ERR("set sensor:%s state:%d error", sensor->dev->name, state);
+		}
+		LOG_INF("%s, sensor:%s ret:%d", __func__, sensor->dev->name, ret);
+	}
+
+	/* initial events */
+	k_mutex_init(&ctx->conn_mutex);
+
+	return ret;
+}
+
+int senss_deinit(void)
+{
+	struct senss_mgmt_context *ctx = get_senss_ctx();
+	struct senss_connection *conn;
+	struct senss_sensor *sensor;
+	int ret = 0;
+	int i;
+
+	/* close all sensors */
+	for (i = 0; i < CONFIG_SENSS_MAX_CONNECTION_COUNT; i++) {
+		conn = ctx->conns[i];
+		if (!conn) {
+			continue;
+		}
+		ret |= close_sensor(conn);
+	}
+
+	/* free sensors */
+	for_each_sensor(ctx, i, sensor) {
+		free(sensor);
+	}
+	ctx->sensor_num = 0;
+	ctx->senss_initialized = false;
+
+	LOG_INF("%s complete", __func__);
```

Comment:
I don't see a free on ctx->info which is malloced in senss_get_sensors. Should it be here, or is somewhere else?

Reply:
remove senss_deinit() now